drivers/i2c/i2c_ite_enhance.c

```diff
@@ -0,0 +1,771 @@
+				/* Send ID */
+				i2c_pio_trans_data(dev, RX_DIRECT,
+					data->addr_16bit << 1, 1);
+			}
+			/* Turn on irq before next direct read */
+			irq_enable(config->i2c_irq_base);
+		} else {
+			if (data->ridx < data->msgs->len) {
+				/* read data */
+				*(data->msgs->buf++) = IT8XXX2_I2C_DRR(base);
+				data->ridx++;
+				/* done */
+				if (data->ridx == data->msgs->len) {
+					data->msgs->len = 0;
+					if (data->msgs->flags & I2C_MSG_STOP) {
+						data->i2ccs = I2C_CH_NORMAL;
+						IT8XXX2_I2C_CTR(base) = E_FINISH;
+						/* wait for stop bit interrupt */
+						data->stop = 1;
+						return 1;
+					}
+					/* End the transaction */
+					data->i2ccs = I2C_CH_WAIT_READ;
+					return 0;
+				}
+				/* read next byte */
+				i2c_pio_trans_data(dev, RX_DIRECT, in_data, 0);
+			}
+		}
+	}
+	return 1;
+}
+
+static int enhanced_i2c_tran_write(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	uint8_t *base = config->base;
+	uint8_t out_data;
+
+	if (data->msgs->flags & I2C_MSG_START) {
+		/* Clear start bit */
+		data->msgs->flags &= ~I2C_MSG_START;
+		enhanced_i2c_start(dev);
+		/* Send ID */
+		i2c_pio_trans_data(dev, TX_DIRECT, data->addr_16bit << 1, 1);
+	} else {
+		/* Host has completed the transmission of a byte */
+		if (data->widx < data->msgs->len) {
+			out_data = *(data->msgs->buf++);
+			data->widx++;
+
+			/* Send Byte */
+			i2c_pio_trans_data(dev, TX_DIRECT, out_data, 0);
+			if (data->i2ccs == I2C_CH_WAIT_NEXT_XFER) {
+				data->i2ccs = I2C_CH_NORMAL;
+				irq_enable(config->i2c_irq_base);
+			}
+		} else {
+			/* done */
+			data->msgs->len = 0;
+			if (data->msgs->flags & I2C_MSG_STOP) {
+				IT8XXX2_I2C_CTR(base) = E_FINISH;
+				/* wait for stop bit interrupt */
+				data->stop = 1;
+			} else {
+				/* Direct write with direct read */
+				data->i2ccs = I2C_CH_WAIT_NEXT_XFER;
+				return 0;
+			}
+		}
+	}
+	return 1;
+}
+
+static int i2c_transaction(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	uint8_t *base = config->base;
+
+	/* no error */
+	if (!(enhanced_i2c_error(dev))) {
+		if (!data->stop) {
+			/* i2c read */
+			if (data->msgs->flags & I2C_MSG_READ) {
+				return enhanced_i2c_tran_read(dev);
+			/* i2c write */
+			} else {
+				return enhanced_i2c_tran_write(dev);
+			}
+		}
+	}
+	/* reset i2c port */
+	i2c_reset(dev);
+	IT8XXX2_I2C_CTR1(base) = 0;
+
+	data->stop = 0;
+	/* done doing work */
+	return 0;
+}
+
+static int i2c_enhance_transfer(const struct device *dev, struct i2c_msg *msgs,
+				uint8_t num_msgs, uint16_t addr)
+{
+	struct i2c_enhance_data *data;
+	const struct i2c_enhance_config *config;
+	int res;
+
+	/* Check for NULL pointers */
+	if (dev == NULL) {
+		LOG_ERR("Device handle is NULL");
+		return -EINVAL;
+	}
+	if (msgs == NULL) {
+		LOG_ERR("Device message is NULL");
+		return -EINVAL;
+	}
```

Comment:
we don't usually make drivers defensive against application programming errors.

Reply:
Removed.